docs/ibc/apps/ibcmodule.md

```diff
@@ -0,0 +1,350 @@
+) (string, error) {
+    // ... do custom initialization logic
+
+    // Use above arguments to determine if we want to abort handshake
+    if err := checkArguments(args); err != nil {
+        return "", err
+    }
+
+    if counterpartyVersion != types.Version {
+		return "", sdkerrors.Wrapf(types.ErrInvalidVersion, "invalid counterparty version: got: %s, expected %s", counterpartyVersion, types.Version)
+	}
+
+    // Module may have already claimed capability in OnChanOpenInit in the case of crossing hellos
+    // (ie chainA and chainB both call ChanOpenInit before one of them calls ChanOpenTry)
+    // If the module can already authenticate the capability then the module already owns it so we don't need to claim
+    // Otherwise, module does not have channel capability and we must claim it from IBC
+    if !im.keeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+        // Only claim channel capability passed back by IBC module if we do not already own it
+        if err := im.keeper.scopedKeeper.ClaimCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)); err != nil {
+            return "", err
+        }
+    }
+
+    return types.Version, nil
+}
+
+// Called by IBC Handler on MsgOpenAck
+func (im IBCModule) OnChanOpenAck(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+    counterpartyVersion string,
+) error {
+    if counterpartyVersion != types.Version {
+		return sdkerrors.Wrapf(types.ErrInvalidVersion, "invalid counterparty version: %s, expected %s", counterpartyVersion, types.Version)
+	}
+
+    // do custom logic
+
+    return nil
+}
+
+// Called by IBC Handler on MsgOpenConfirm
+func (im IBCModule) OnChanOpenConfirm(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+) error {
+    // do custom logic
+
+    return nil
+}
+```
+
+The channel closing handshake will also invoke module callbacks that can return errors to abort the closing handshake. Closing a channel is a 2-step handshake, the initiating chain calls `ChanCloseInit` and the finalizing chain calls `ChanCloseConfirm`.
+
+```go
+// Called by IBC Handler on MsgCloseInit
+func (im IBCModule) OnChanCloseInit(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+) error {
+    // ... do custom finalization logic
+
+    // Use above arguments to determine if we want to abort handshake
+    err := checkArguments(args)
+    return err
+}
+
+// Called by IBC Handler on MsgCloseConfirm
+func (im IBCModule) OnChanCloseConfirm(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+) error {
+    // ... do custom finalization logic
+
+    // Use above arguments to determine if we want to abort handshake
+    err := checkArguments(args)
+    return err
+}
+```
+
+### Channel handshake version negotiation
+
+Application modules are expected to verify versioning used during the channel handshake procedure.
+
+- `onChanOpenInit` will verify that the relayer-chosen parameters
```

Comment:
Nit: The `o` in all the `on...` callbacks that you go through here is lowercase but in the code sample above the `o` is uppercase. Maybe make it uppercase here as well?